### ISSUE TITLE
DiagnosticSource.StartActivity and DiagnosticSource.StopActivity allocate a string per call

### ISSUE THREAD

**Issue by davidfowl** [dotnet/corefx#37055]
StartActivity

https://github.com/dotnet/corefx/blob/43ca5e2aee8c4ada93ab30fc3fd34bb8f4c6224d/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceActivity.cs#L31

StopActivity

https://github.com/dotnet/corefx/blob/43ca5e2aee8c4ada93ab30fc3fd34bb8f4c6224d/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceActivity.cs#L51

Doing some profiling with acitivities on it turns out we end up allocating per call to Stop and Start (a needless allocation as well). The caller can simply pass in a string that has ".Start" and ".Stop" appeneded to avoid the string concatenation:

Here's a profile memory profile of 100K requests and it shows the backtraces from StopActivity (StartActivity isn't being called here).

![image](https://user-images.githubusercontent.com/95136/56454356-2ac3ba00-6304-11e9-81ae-5a651f55454d.png)

**Comment by Wraith2** [comment]
Looks like the obvious solution is as you suggest to get the caller to put in the .Start and .Stop suffixes, but that's a breaking change and I don't know if tools like perfview use those suffixes to detect start/stop for aggregation.

How about check the input name for ending in exactly the right suffix using an ordinal comparison. If it is pre-suffixed just pass it through otherwise use the existing behaviour. 
It'd be nice to put a cache in there so existing users get some benefit but that would lead to having to decide what to cache for how long etc.

**Comment by davidfowl** [comment]
> Looks like the obvious solution is as you suggest to get the caller to put in the .Start and .Stop suffixes, but that's a breaking change and I don't know if tools like perfview use those suffixes to detect start/stop for aggregation so we might not just be able to change to not using them.

Right, we can't change the current name.

> How about check the input name for ending in exactly the right suffix using an ordinal comparison. If it is pre-suffixed just pass it through otherwise use the existing behaviour.

I'm not sure that would work since the name for starting and stopping are different. The operation name is tied to the activity instance.

> It'd be nice to put a cache in there so existing users get some benefit but that would lead to having to decide what to cache for how long etc.

It's attractive but its basically a memory leak. Maybe it could be a fixed size cache? 

A few options:
- Add the cache as you suggest (fixed size)
- Add overloads that let the caller pass in the start/stop operation name

The latter seems like it would make the most sense but existing producers won't get the benefit.

**Comment by benaadams** [comment]
Something like?
```csharp
public Activity StartActivity(Activity activity, object args)
{
    activity.Start();
    string operationName = activity.OperationName;
    if (!operationName.EndsWith(".Start", StringComparison.OrdinalIgnoreCase))
    {
        operationName += ".Start"
    }
    Write(operationName, args);
    return activity;
}
```

**Comment by Wraith2** [comment]
Yes, that's what I had in mind. Any idea on a sane fixed size cache?

**Comment by davidfowl** [comment]
That check won’t help unless we’re going to start creating different activities for start and stop

**Comment by Wraith2** [comment]
Then the cache is the way to go because it's not just being written to something it becomes part of a KeyValuePair so it needs to be real.

**Comment by davidfowl** [comment]
Honestly I’d rather add overloads and update the ~10 or so places that use this APi

**Comment by stephentoub** [comment]
Activity looks relatively heavy-weight in terms of allocations, so presumably if these allocations matter it's because Start/Stop are used many times per Activity? If so, why not just store the lazily-initialized +.Start and +.Stop as fields internally on the Activity and have it use those repeatedly?

>  I’d rather add overloads 

That do what? Ultimately won't they need to call the same abstract write method that takes the singles string name?

I'm partially confused by this issue because as DiagnosticSource was being built I was pushing on the costs involved, and I thought I was told it was used in situations where the allocation overheads didn't matter. Is that not the case?

**Comment by davidfowl** [comment]
> Activity looks relatively heavy-weight in terms of allocations

This is taking baby steps to fixing that. I just started profiling this thing (we're looking at the performance more now because they are effectively on by default in ASP.NET Core in 3.0).

> So presumably if these allocations matter it's because Start/Stop are used many times per Activity? If so, why not just store the lazily-initialized +.Start and +.Stop as fields internally on the Activity and have it use those repeatedly?

True, we could just add some internal fields to the Activity itself (for some reason I was thinking the cache needed to be separate).

> That do what? Ultimately won't they need to call the same abstract write method that takes the singles string name?

Overloads that take the name directly. But we can just cache 2 more fields on the activity itself (though that'll make it bigger 😄).

> I'm partially confused by this issue because as DiagnosticSource was being built I was pushing on the costs involved, and I thought I was told it was used in situations where the allocation overheads didn't matter. Is that not the case?

Yes that isn't the case, they totally matter. Before it was less so because the scenarios where they were being used there were other costs that maybe over shadowed the cost of the Activity. Now that we'tr trying to make distributed tracing be as "on by default" as possible these are starting to show up.

**Comment by Wraith2** [comment]
How about 
```csharp
 public Activity StartActivity(Activity activity, Tuple<string,string> startAndStopNames, object args)
```
to avoid existing uses where the second param is already a string compiling and then failing. Then just store a ref to the tuple and null check it on start and stop and take the value if present else fallback to current.

**Comment by davidfowl** [comment]
I'm likely going to change ASP.NET Core to remove these calls and replicate what they are doing just passing in the right names. It avoids bloating activity further, and the strings can be static/const.
